# Hierarchical Clustering using Euclidean Distance  
# Task 1: Introduction  
## - Extending Skew Analysis

Six skews of different combinations of two nucleotides: CA-, GA-, UA-, UG-, UC-, and CG-skew are used to draw what is known as the **skew profile**. The skew profile is plotted using **cumulative** values of skews and is determined by nucleotide **composition**, not the sequence. Despite the benefits of this method for knowing the genome profile, one of its flaws is the lack of a quantitative comparison between more than one genome. This is being overcome by additional techniques such as [Euclidean distance](https://en.wikipedia.org/wiki/Euclidean_distance) matrices and [Neighbor-joining](https://en.wikipedia.org/wiki/Neighbor_joining) trees. Our traget in the current project is to explore these options.  


The following twenty skew profile graphs represent different strains for seven viruses, Corona and SARS viruses from the [Coronaviridae](https://en.wikipedia.org/wiki/Coronaviridae) family, Dengue, Zika, and West Nile viruses from the [Flaviviridae](https://en.wikipedia.org/wiki/Flaviviridae) family, Enterovirus from [Picornaviridae](https://en.wikipedia.org/wiki/Picornavirus) family, and HIV from [Retroviridae](https://en.wikipedia.org/w/index.php?title=Retroviridae&redirect=yes). All these viruses belong to Realm [Riboviria](https://en.wikipedia.org/wiki/Riboviria).  

<table><tr><td><img src='images/Corona_HCoV-NL63.png'></td><td><img src='images/Corona_MN988668_China.png'></td></tr></table>
<table><tr><td><img src='images/Corona_MT755827_Bangladesh.png'></td><td><img src='images/Corona_MT759582.1_India.png'></td></tr></table>
<table><tr><td><img src='images/Corona_MT766907.1_USA.png'></td><td><img src='images/Corona_NC_045512_China.png'></td></tr></table>
<table><tr><td><img src='images/SARS_BJ01.png'></td><td><img src='images/SARS_BJ02.png'></td></tr></table>
<table><tr><td><img src='images/SARS_BJ03.png'></td><td><img src='images/SARS_BJ04.png'></td></tr></table>
<table><tr><td><img src='images/Dengue_MT862858.png'></td><td><img src='images/Dengue_MT862893.png'></td></tr></table>

<table><tr><td><img src='images/EnterovirusA_NC038306.png'></td><td><img src='images/EnterovirusB_NC038307.png'></td></tr></table>
<table><tr><td><img src='images/EnterovirusD_NC038308.png'></td><td><img src='images/EnterovirusH_NC038309.png'></td></tr></table>
<table><tr><td><img src='images/HIV_NC001802.1.png'></td><td><img src='images/WestNile_NC009942.png'></td></tr></table>

<table><tr><td><img src='images/Zika_AY632535.png'></td><td><img src='images/Zika_MN101548.png'></td></tr></table>


If you've tried to compare the previous 20 graphs with each other, you know how difficult it can be to accomplish that with sight. We will accomplish that by starting with the profile skew data, like the one shown in the next table, and compute the Euclidean differences among them and build a [dendrogram](https://en.wikipedia.org/wiki/Dendrogram) based on the neighbor-joining tree algorithm, which shows the difference and similarity between the different viruses.   

No. | virus | strain |CA-Skew | GA-Skew | UA-Skew | UG-Skew | UC-Skew | CG-Skew*  
----| :-------- | :-- | -- | -- | -- | -- |----------| ---------
1. | **Corona** | HCoV-NL63 |-8892.72|-13454.38|-5947.28|8392.05|3165.32 |5417.59  
2. | **SARS** | BJ01 |-4673.66|-5852.78|-1008.29|4887.03|3697.07 |1216.47  
3. | **Dengue** | MT862858 |816.63|-351.01|1728.46|2067.33|926.05 |1164.02  
4. | **Enterovirus A** | NC038306 |33.56|201.33|429.91|229.41|397.55 |-165.63  
5. | **HIV** | NC001802 |447.21|-939.66|2171.98|3040.84|1749.07 |1384.04  
6. | **West Nile** | NC009942 |1324.54|-57.23|1054.20|1111.13|-273.42 |1382.03  
7. | **Zika** | AY632535 |1381.19|-283.89|1147.27|1432.18|-237.85 |1663.35  


  *It should be noted that, in skew language, CG does not represent a CG base pair but a comparison of the C with the G nucleotide proportions.*   

Our target in this project is to create a dendrogram like the next dendrograms created using [**MEGA** *X* (**M**olecular **E**volutionary **G**enetics **A**nalysis software )](https://www.megasoftware.net/). Please note that this dendrogram was constructed by relying on [pairwise distances](https://en.wikipedia.org/wiki/Distance_matrix) and [multiple sequence alignments](https://en.wikipedia.org/wiki/Multiple_sequence_alignment). The dendrogram that we will create will depend on the cumulative skew profile, which in turn depends on the nucleotide **composition**, not the sequence.  
<table><tr><td><img src='images/dendogram_H.png'></td><td><img src='images/dendogram.png'></td></tr></table>  

Dendrogram (1) |  | Dendrogram (2) |
---- | |-------- |  
**Vertical lines show the distances** |  | **Horizontal lines show the distances** |

The dendrogram can be drawn in any direction. In dendrogram (1), the lengths of vertical lines show the distances between different strains, while the lengths of the horizontal lines in this form of tree do not signify anything. They are merely drawn to space the strains conveniently. The opposite is in the dendrogram (2), the horizontal lines show the distances.  


# Task 2: 
## 2.1. Importing Libraries 

First of all, we will need to import some libraries. These include 

- [os](https://docs.python.org/3/library/os.html) - Miscellaneous operating system interfaces,  
- [statistics](https://docs.python.org/3/library/statistics.html) - Mathematical statistics functions  
- [numpy](https://numpy.org/) - Package for scientific computing,  
- [pandas](https://pandas.pydata.org/) - data analysis and manipulation tool,
- [matplotlib](https://matplotlib.org/users/index.html) - Visualization with Python,
    - [pyplot](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.html#module-matplotlib.pyplot) - Interactive plots (*MATLAB-like way of plotting*),  
- [scipy](https://www.scipy.org/) - Python-based ecosystem of open-source software for mathematics, science, and engineering. 
    - [dendrogram](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.dendrogram.html)  
    - [linkage](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html)  



In [1]:
import os
import statistics
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
from scipy.cluster.hierarchy import dendrogram, linkage

##  2.2. Locate the Data Files

All viruses that we will be analyzing are [RNA viruses](https://en.wikipedia.org/wiki/RNA_virus). In contrast to [DNA viruses](https://en.wikipedia.org/wiki/DNA_virus), which can reach **300** kilobases (kb) in size, RNA viruses have a size of about 30 kb. In late 2018, the [planarian nidovirus](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6211748/) genome size of **41** kb was discovered seting a new recored for the length of RNA genomes.  
   
**Classification & Genome sizes**:    
  - Riboviria  
     - Coronaviridae family,  
      - [Coronavirus](https://en.wikipedia.org/wiki/Severe_acute_respiratory_syndrome_coronavirus_2) ≈ 30 Kb  
      - [SARS Coronavirus](https://en.wikipedia.org/wiki/Severe_acute_respiratory_syndrome_coronavirus) ≈ 30 Kb  
     - Flaviviridae family,  
        - [Dengue virus](https://en.wikipedia.org/wiki/Dengue_virus) ≈ 11 Kb   
        - [Zika](https://en.wikipedia.org/wiki/Zika_virus) ≈ 11 Kb  
        - [West Nile virus](https://en.wikipedia.org/wiki/West_Nile_virus) ≈ 11 Kb  
     - Picornaviridae  
        - [Enterovirus](https://en.wikipedia.org/wiki/Enterovirus) ≈ 8 Kb   
     - Retroviridae.  
        - [HIV](https://en.wikipedia.org/wiki/HIV) ≈ 10 Kb   

 *N.B. Keep an eye on the [Corona_HCoV-NL63](https://en.wikipedia.org/wiki/Human_coronavirus_NL63) strain. It is one of the first strains discovered back in 2004 (doi: [10.1038/nm1024. Epub 2004 Mar 21](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7095789/)). It is interesting to know the differences and similarities between this strain and the newly discovered ones.*

In [2]:
# List all data files 
# You need to change the path
data_path = 'C:/Users/Administrator/Documents/data'
for file in os.listdir(data_path):
    print(file)

Corona_HCoV-NL63.txt
Corona_MN988668.txt
Corona_MT755827.txt
Corona_MT759582.txt
Corona_MT766907.txt
Corona_NC045512.txt
Dengue_MT862858.txt
Dengue_MT862893.txt
Enterovirus A_NC038306.txt
Enterovirus B_NC038307.txt
Enterovirus D_NC038308.txt
Enterovirus H_NC038309.txt
HIV_NC001802.txt
SARS_BJ01.txt
SARS_BJ02.txt
SARS_BJ03.txt
SARS_BJ04.txt
West Nile_NC009942.txt
Zika_AY632535.txt
Zika_MN101548.txt


##  2.3.  Calculate the Cumulative and Mean Skew Values.


In [3]:
# define 'bases_skew' function to calculate the skew values.
def bases_skew(A, B):
    try: return (A - B) / (A + B)
    except ZeroDivisionError: return 0

In [4]:
mat = np.array([])                # Cumulative skew values
mat2 = np.array([])               # Mean skew values
virus_names = list()

for file in os.listdir(data_path):
    input_file_path = data_path + '/' + file
    #print(input_file_path)
    counter = 0; A_count = 0; C_count = 0; G_count = 0; U_count = 0
    ca_skew = []; ga_skew = []; ua_skew = []; uc_skew = []; ug_skew = []; cg_skew = []  
    
    temp_DNA = ''           # One line of Template DNA Sequence
    with open(input_file_path,'r') as input_data:
        header = input_data.readline().strip()
        for line in input_data:
            temp_DNA = line.strip()
            for base in temp_DNA:
                counter += 1
                if base == "A":
                    U_count +=1                  
                elif base == "C":
                    G_count +=1
                elif base == "G":
                    C_count +=1
                elif base == "T":
                    A_count +=1        

                ca_skew.insert(counter, bases_skew(C_count, A_count))
                ga_skew.insert(counter, bases_skew(G_count, A_count))
                ua_skew.insert(counter, bases_skew(U_count, A_count))
                ug_skew.insert(counter, bases_skew(U_count, G_count))
                uc_skew.insert(counter, bases_skew(U_count, C_count))
                cg_skew.insert(counter, bases_skew(C_count, G_count))            

        #print('File name ', file)
        #print('Total bases: ', counter)
        #print('Cumulative ca_skew = ', np.cumsum(ca_skew)[len(ca_skew)-1])
        #print('Cumulative ga_skew = ',np.cumsum(ga_skew)[len(ga_skew)-1])
        #print('Cumulative ua_skew = ',np.cumsum(ua_skew)[len(ua_skew)-1])
        #print('Cumulative ug_skew = ',np.cumsum(ug_skew)[len(ug_skew)-1])
        #print('Cumulative uc_skew = ',np.cumsum(uc_skew)[len(uc_skew)-1])
        #print('Cumulative cg_skew = ',np.cumsum(cg_skew)[len(cg_skew)-1])
        #print('==============================================')
        #print('Mean ca_skew = ', statistics.mean(ca_skew))
        #print('Mean ga_skew = ', statistics.mean(ga_skew))
        #print('Mean ua_skew = ', statistics.mean(ua_skew))
        #print('Mean ug_skew = ', statistics.mean(ug_skew))
        #print('Mean uc_skew = ', statistics.mean(uc_skew))
        #print('Mean cg_skew = ', statistics.mean(cg_skew))
        
        # Get the virus name from the file name
        virus_Name = os.path.split(input_file_path)[1].split(".")[0]
        # Insert the virus name into the virus names list
        virus_names.append(virus_Name)
        c = 10
        if mat.shape == (0,):
            #print('The mat is empty, use hstack.')
            mat = np.hstack((mat, np.array([np.cumsum(ca_skew)[len(ca_skew)-1],
                                            np.cumsum(ga_skew)[len(ga_skew)-1],
                                            np.cumsum(ua_skew)[len(ua_skew)-1],
                                            np.cumsum(ug_skew)[len(ug_skew)-1],
                                            np.cumsum(uc_skew)[len(uc_skew)-1],
                                            np.cumsum(cg_skew)[len(cg_skew)-1]                                            
                                           ])))
            mat2 = np.hstack((mat2, np.array([(statistics.mean(ca_skew))*c,
                                            (statistics.mean(ga_skew))*c,
                                            (statistics.mean(ua_skew))*c,
                                            (statistics.mean(ug_skew))*c,
                                            (statistics.mean(uc_skew))*c,
                                            (statistics.mean(cg_skew))*c                                            
                                           ])))
        else:
            #print('The mat has at least one row, use vstack.')
            mat = np.vstack((mat, np.array([np.cumsum(ca_skew)[len(ca_skew)-1],
                                            np.cumsum(ga_skew)[len(ga_skew)-1],
                                            np.cumsum(ua_skew)[len(ua_skew)-1],
                                            np.cumsum(ug_skew)[len(ug_skew)-1],
                                            np.cumsum(uc_skew)[len(uc_skew)-1],
                                            np.cumsum(cg_skew)[len(cg_skew)-1]                                            
                                           ])))    
            mat2 = np.vstack((mat2, np.array([(statistics.mean(ca_skew))*c,
                                            (statistics.mean(ga_skew))*c,
                                            (statistics.mean(ua_skew))*c,
                                            (statistics.mean(ug_skew))*c,
                                            (statistics.mean(uc_skew))*c,
                                            (statistics.mean(cg_skew))*c                                            
                                           ])))    
        
        #plt.figure(figsize=(9,6))
        #plt.plot(np.cumsum(ca_skew), label="Cumulative CA Skew") 
        #plt.plot(np.cumsum(ga_skew), label="Cumulative GA Skew") 
        #plt.plot(np.cumsum(ua_skew), label="Cumulative UA Skew") 
        #plt.plot(np.cumsum(ug_skew), label="Cumulative UG Skew") 
        #plt.plot(np.cumsum(uc_skew), label="Cumulative UC Skew") 
        #plt.plot(np.cumsum(cg_skew), label="Cumulative CG Skew")      
        #plt.title(virus_Name + " Skew Profiles")
        #plt.legend()
        #plt.grid()
        ### You can use 'plt.savefig()' to save the plot, but that will slow down the code even more!
        ##plt.savefig(virus_Name + '.png', dpi=72, bbox_inches='tight')
        #plt.show()    
        #plt.close()
        input_data.close()

In [5]:
linkage(mat2)[0,]

array([1.       , 5.       , 0.0096476, 2.       ])

# Task 3: Understand the data set

In [6]:
print(len(virus_names))
virus_names[:]

20


['Corona_HCoV-NL63',
 'Corona_MN988668',
 'Corona_MT755827',
 'Corona_MT759582',
 'Corona_MT766907',
 'Corona_NC045512',
 'Dengue_MT862858',
 'Dengue_MT862893',
 'Enterovirus A_NC038306',
 'Enterovirus B_NC038307',
 'Enterovirus D_NC038308',
 'Enterovirus H_NC038309',
 'HIV_NC001802',
 'SARS_BJ01',
 'SARS_BJ02',
 'SARS_BJ03',
 'SARS_BJ04',
 'West Nile_NC009942',
 'Zika_AY632535',
 'Zika_MN101548']

In [7]:
print('mat shape = ', mat.shape)
print('mat2 shape = ', mat2.shape)

mat shape =  (20, 6)
mat2 shape =  (20, 6)


In [8]:
print(mat[0,0])
print("{:.2f}".format(mat[0,0]))

-8892.718135634901
-8892.72


In [9]:
print(mat[0,])

[ -8892.71813563 -13454.3840888   -5947.28102079   8392.04623687
   3165.31804809   5417.58910545]


In [10]:
np.set_printoptions(linewidth=150)
print('0 ', mat[0,])
print('1 ', mat[1,])
print('2 ', mat[2,])
print('...', '...')
print('...', '...')
print('19 ', mat[19,])

0  [ -8892.71813563 -13454.3840888   -5947.28102079   8392.04623687   3165.31804809   5417.58910545]
1  [-6452.98699289 -8028.54945613  -918.9649018   7170.21888199  5574.67104164  1676.44382901]
2  [-6476.32615371 -8071.24025314  -928.13283219  7203.71647803  5588.47631333  1697.3942333 ]
... ...
... ...
19  [1352.1148887  -295.92443703 1152.42689342 1444.8624689  -202.88952768 1644.59149638]


In [11]:
print('0 ', mat2[0,])
print('1 ', mat2[1,])
print('2 ', mat2[2,])
print('...', '...')
print('...', '...')
print('19 ', mat2[19,])

0  [-3.22749542 -4.88309225 -2.15848765  3.04578312  1.14881067  1.96624292]
1  [-2.15956193 -2.68684095 -0.30754155  2.39959134  1.86562399  0.56104007]
2  [-2.16577807 -2.69914064 -0.31038118  2.40902802  1.86886811  0.56763343]
... ...
... ...
19  [ 1.2591869  -0.27558618  1.07322303  1.34556013 -0.18894536  1.53156221]


In [12]:
print(list(map('{:.1f}'.format, mat[0,])))
print(list(map('{:.1f}'.format, mat[1,])))
print(list(map('{:.1f}'.format, mat[2,])))
print(list(map('{:.1f}'.format, mat[3,])))

['-8892.7', '-13454.4', '-5947.3', '8392.0', '3165.3', '5417.6']
['-6453.0', '-8028.5', '-919.0', '7170.2', '5574.7', '1676.4']
['-6476.3', '-8071.2', '-928.1', '7203.7', '5588.5', '1697.4']
['-6392.6', '-8084.7', '-1013.4', '7133.5', '5420.5', '1795.3']


In [13]:
print(list(map('{:.3f}'.format, [i/10 for i in mat2[0,]])))
print(list(map('{:.3f}'.format, [i/10 for i in mat2[1,]])))
print(list(map('{:.3f}'.format, [i/10 for i in mat2[2,]])))
print('...', '...')
print('...', '...')
print(list(map('{:.3f}'.format, [i/10 for i in mat2[19,]])))

['-0.323', '-0.488', '-0.216', '0.305', '0.115', '0.197']
['-0.216', '-0.269', '-0.031', '0.240', '0.187', '0.056']
['-0.217', '-0.270', '-0.031', '0.241', '0.187', '0.057']
... ...
... ...
['0.126', '-0.028', '0.107', '0.135', '-0.019', '0.153']


In [14]:
df = pd.DataFrame(np.array([
        list(map('{:.2f}'.format, mat[0,])), list(map('{:.2f}'.format, mat[1,])),
        list(map('{:.2f}'.format, mat[2,])), list(map('{:.2f}'.format, mat[3,])),
        list(map('{:.2f}'.format, mat[4,])), list(map('{:.2f}'.format, mat[5,])),
        list(map('{:.2f}'.format, mat[6,])), list(map('{:.2f}'.format, mat[7,])),
        list(map('{:.2f}'.format, mat[8,])), list(map('{:.2f}'.format, mat[9,])),
        list(map('{:.2f}'.format, mat[10,])), list(map('{:.2f}'.format, mat[11,])),
        list(map('{:.2f}'.format, mat[12,])), list(map('{:.2f}'.format, mat[13,])),
        list(map('{:.2f}'.format, mat[14,])), list(map('{:.2f}'.format, mat[15,])),
        list(map('{:.2f}'.format, mat[16,])), list(map('{:.2f}'.format, mat[17,])),
        list(map('{:.2f}'.format, mat[18,])), list(map('{:.2f}'.format, mat[19,]))      
        ]),
        columns=['ca_skew', 'ga_skew', 'ua_skew','ug_skew', 'uc_skew', 'cg_skew'],
        index=virus_names[:])
df

ca_skew    ga_skew   ua_skew  ug_skew  uc_skew  \
Corona_HCoV-NL63        -8892.72  -13454.38  -5947.28  8392.05  3165.32   
Corona_MN988668         -6452.99   -8028.55   -918.96  7170.22  5574.67   
Corona_MT755827         -6476.33   -8071.24   -928.13  7203.72  5588.48   
Corona_MT759582         -6392.58   -8084.69  -1013.36  7133.52  5420.50   
Corona_MT766907         -6347.71   -8089.15   -943.60  7206.93  5444.58   
Corona_NC045512         -6458.06   -8034.26   -902.06  7191.16  5595.44   
Dengue_MT862858           907.99    -114.88   1841.00  1955.26   952.23   
Dengue_MT862893           917.53    -124.85   1845.74  1969.50   947.53   
Enterovirus A_NC038306      7.85     188.96    270.06    80.26   260.78   
Enterovirus B_NC038307    134.34     201.25    596.30   395.27   461.43   
Enterovirus D_NC038308   -842.53    -747.48    381.86  1118.93  1212.60   
Enterovirus H_NC038309   -255.04    -298.53    290.58   586.04   543.55   
HIV_NC001802              662.92    -648.06   2278.82  2865.17  1644.05   
SARS_BJ01               -4673.66   -5852.78  -1008.29  4887.03  3697.07   
SARS_BJ02               -4795.84   -5961.80  -1080.74  4924.53  3742.74   
SARS_BJ03               -4758.62   -5908.16  -1096.24  4856.45  3689.25   
SARS_BJ04               -4711.67   -5868.52  -1017.46  4892.61  3722.20   
West Nile_NC009942       1324.54     -57.23   1054.20  1111.13  -273.42   
Zika_AY632535            1381.19    -283.89   1147.27  1432.18  -237.85   
Zika_MN101548            1352.11    -295.92   1152.43  1444.86  -202.89   

                        cg_skew  
Corona_HCoV-NL63        5417.59  
Corona_MN988668         1676.44  
Corona_MT755827         1697.39  
Corona_MT759582         1795.32  
Corona_MT766907         1846.75  
Corona_NC045512         1677.13  
Dengue_MT862858         1021.56  
Dengue_MT862893         1040.92  
Enterovirus A_NC038306  -182.11  
Enterovirus B_NC038307   -67.37  
Enterovirus D_NC038308   -96.39  
Enterovirus H_NC038309    44.60  
HIV_NC001802            1306.64  
SARS_BJ01               1216.47  
SARS_BJ02               1209.34  
SARS_BJ03               1190.51  
SARS_BJ04               1197.49  
West Nile_NC009942      1382.03  
Zika_AY632535           1663.35  
Zika_MN101548           1644.59

In [15]:
df2 = pd.DataFrame(np.array([
        list(map('{:.2f}'.format, [i/10 for i in mat2[0,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[1,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[2,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[3,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[4,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[5,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[6,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[7,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[8,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[9,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[10,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[11,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[12,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[13,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[14,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[15,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[16,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[17,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[18,]])), 
        list(map('{:.2f}'.format, [i/10 for i in mat2[19,]]))    
        ]),
        columns=['ca_skew', 'ga_skew', 'ua_skew','ug_skew', 'uc_skew', 'cg_skew'],
        index=virus_names[:])
df2

ca_skew ga_skew ua_skew ug_skew uc_skew cg_skew
Corona_HCoV-NL63         -0.32   -0.49   -0.22    0.30    0.11    0.20
Corona_MN988668          -0.22   -0.27   -0.03    0.24    0.19    0.06
Corona_MT755827          -0.22   -0.27   -0.03    0.24    0.19    0.06
Corona_MT759582          -0.21   -0.27   -0.03    0.24    0.18    0.06
Corona_MT766907          -0.21   -0.27   -0.03    0.24    0.18    0.06
Corona_NC045512          -0.22   -0.27   -0.03    0.24    0.19    0.06
Dengue_MT862858           0.09   -0.01    0.18    0.19    0.09    0.10
Dengue_MT862893           0.09   -0.01    0.18    0.19    0.09    0.10
Enterovirus A_NC038306    0.00    0.03    0.04    0.01    0.04   -0.02
Enterovirus B_NC038307    0.02    0.03    0.08    0.05    0.06   -0.01
Enterovirus D_NC038308   -0.11   -0.10    0.05    0.15    0.16   -0.01
Enterovirus H_NC038309   -0.03   -0.04    0.04    0.08    0.07    0.01
HIV_NC001802              0.07   -0.07    0.25    0.31    0.18    0.14
SARS_BJ01                -0.16   -0.20   -0.03    0.16    0.12    0.04
SARS_BJ02                -0.16   -0.20   -0.04    0.17    0.13    0.04
SARS_BJ03                -0.16   -0.20   -0.04    0.16    0.12    0.04
SARS_BJ04                -0.16   -0.20   -0.03    0.16    0.13    0.04
West Nile_NC009942        0.12   -0.01    0.10    0.10   -0.02    0.13
Zika_AY632535             0.13   -0.03    0.11    0.13   -0.02    0.15
Zika_MN101548             0.13   -0.03    0.11    0.13   -0.02    0.15

# Task 4: Hierarchical Clustering - Metric

> Wikipedia
>> In [Cartesian coordinates](https://en.wikipedia.org/wiki/Cartesian_coordinate_system) if $p = (p_1, p_2,\cdots, p_n)$ and $q = (q_1, q_2,\cdots, q_n)$ are two points in [Euclidean n-space](https://en.wikipedia.org/wiki/Euclidean_space), then the **Euclidean distance** (d) from **p** to **q**, or from **q** to **p**, is given by the [Pythagorean formula](https://en.wikipedia.org/wiki/Pythagorean_theorem):  
>>$d(p, q) = d(q, p)= \sqrt{\sum_{i=1}^n (q_i - p_i)^2}$  
>>$=\sqrt{(q_1-p_1)^2+(q_2-p_2)^2+\cdots+(q_n-p_n)^2}$  
  
  
$d(Zika AY632535, Zika MN101548)=
\sqrt{(1381.19-1352.11)^2+(-283.89-(-295.92))^2+(1147.27-1152.43)^2+
(1432.18-1444.86)^2+(-237.85-(-202.89))^2+(1663.35-1644.59)^2}\approx52.46$




>```python
import math  
dist = 0  
for p, q in zip(mat[18,], mat[19,]):        
    dist += math.pow((p-q), 2)  
dist = math.sqrt(dist)
print(dist)  
> 52.45662363913553  
>```
  
```python
from scipy.cluster.hierarchy import linkage
linkage(mat)
```

In [16]:
linkage_matrix = linkage(mat, 
                         method='single', 
                         metric='euclidean', 
                         optimal_ordering=False)
print(linkage_matrix.shape)
print(linkage_matrix[0:3])
print('...', '...')
print(linkage_matrix[18])

(19, 4)
[[ 6.          7.         28.5101875   2.        ]
 [ 1.          5.         34.84341166  2.        ]
 [18.         19.         52.45662364  2.        ]]
... ...
[   0.           37.         8900.47217989   20.        ]


In [17]:
y = 0
for x in range(len(linkage_matrix)):
    y+=1
    print(len(linkage_matrix)+y, linkage_matrix[x])

20 [ 6.         7.        28.5101875  2.       ]
21 [ 1.          5.         34.84341166  2.        ]
22 [18.         19.         52.45662364  2.        ]
23 [13.         16.         52.91056314  2.        ]
24 [ 2.        21.        54.7541342  3.       ]
25 [ 14.          15.         111.14546016   2.        ]
26 [ 23.          25.         111.45526329   4.        ]
27 [  3.           4.         124.55923965   2.        ]
28 [ 24.          27.         234.43283633   5.        ]
29 [ 17.          22.         496.71392088   3.        ]
30 [  8.           9.         524.63500631   2.        ]
31 [ 11.          30.         741.97419156   3.        ]
32 [  10.           31.         1143.03514609    4.        ]
33 [  12.           20.         1372.62989182    3.        ]
34 [  29.           33.         1631.54140784    6.        ]
35 [  32.           34.         2201.42550557   10.        ]
36 [  26.           28.         3885.69989123    9.        ]
37 [  35.          36.        8047.9124

In [18]:
df.iloc[[2, 1,5], :]

ca_skew   ga_skew  ua_skew  ug_skew  uc_skew  cg_skew
Corona_MT755827  -6476.33  -8071.24  -928.13  7203.72  5588.48  1697.39
Corona_MN988668  -6452.99  -8028.55  -918.96  7170.22  5574.67  1676.44
Corona_NC045512  -6458.06  -8034.26  -902.06  7191.16  5595.44  1677.13

# Task 5: Hierarchical Clustering - Ordering & Methods

## 5.1: Optimal Ordering

> **Higgs and Attwood (2005)**:
>>*It is often said that trees (the phylogenetics trees) are like hanging mobiles. You can imagine suspending them from the root and allowing the horizontal lines to swing around.*  


<table><tr><td><img src='images/p_tree_1.png'></td><td><img src='images/p_tree_2.png'></td></tr></table>  

Dendrogram (a) |  | Dendrogram (b) |
---- | |:-------- |  
**Vertical lines show the distances** |  | **Vertical lines show the distances** |

Rooted trees. Tree (a) can be converted to tree (b) by swinging around the horizontal branches like mobiles. Hence (a) and
(b) are equivalent to one another. (*modified from Higgs and Attwood - 2005*)


**Paul G. Higgs and Teresa K. Attwood** (2005). Bioinformattics and Molecular Evolution. Willey-Blackwell. Print ISBN:9781405106832 |Online ISBN:9781118697078 |[DOI:10.1002/9781118697078](https://onlinelibrary.wiley.com/doi/book/10.1002/9781118697078)


In [19]:
linkage_matrix = linkage(mat, 
                         method='single', 
                         metric='euclidean', 
                         optimal_ordering=True)
print(linkage_matrix.shape)
for x in range(len(linkage_matrix)):
    print(linkage_matrix[x])

(19, 4)
[ 6.         7.        28.5101875  2.       ]
[ 5.          1.         34.84341166  2.        ]
[18.         19.         52.45662364  2.        ]
[16.         13.         52.91056314  2.        ]
[ 2.        21.        54.7541342  3.       ]
[ 14.          15.         111.14546016   2.        ]
[ 25.          23.         111.45526329   4.        ]
[  3.           4.         124.55923965   2.        ]
[ 24.          27.         234.43283633   5.        ]
[ 17.          22.         496.71392088   3.        ]
[  9.           8.         524.63500631   2.        ]
[ 11.          30.         741.97419156   3.        ]
[  10.           31.         1143.03514609    4.        ]
[  20.           12.         1372.62989182    3.        ]
[  29.           33.         1631.54140784    6.        ]
[  32.           34.         2201.42550557   10.        ]
[  28.           26.         3885.69989123    9.        ]
[  36.          35.        8047.9124615   19.       ]
[   0.           37.        

## 5.2: Clustring Methods

```Python
methods = ('single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward')
```

- **single** $^a$  
  - $d(u,v) = min(dist(u[i], v[j])$  
   - [Nearest Point Algorithm](https://en.wikipedia.org/wiki/Nearest_neighbor_search)  
  
- **complete** $^a$  
  - $d(u,v) = max(dist(u[i], v[j])$  
   - [Complete-linkage clustering](https://en.wikipedia.org/wiki/Complete-linkage_clustering) known also as Farthest Point Algorithm or Voor Hees Algorithm  
  
  **average** $^b$  
   - $d(u,v) = \sum_{ij}\frac{d(u[i], v[j])}{(|u| * |v|)}$  
    - [UPGMA](https://en.wikipedia.org/wiki/UPGMA) **U**nweighted **P**air **G**roup **M**ethod with **A**rithmetic Mean algorithm.   
  
- **weighted** $^c$    
  - $d(u,v) = \frac{dist(s,v) + dist(t,v)}{2}$  
   - [WPGMA](https://en.wikipedia.org/wiki/WPGMA) (**W**eighted **P**air **G**roup **M**ethod with **A**rithmetic Mean)
  
- **centroid** $^d$ 
  - $dist(s,t) = ||c_s - c_t||2$ 
   - [UPGMC](https://en.wikipedia.org/wiki/Hierarchical_clustering) (**U**nweighted **P**air **G**roup **M**ethod with **C**entroid linkage)
- **median**  
  - $d{(i\cup j),k} = \frac{d_{i,k} + d_{j,k}}{2}$ 
   - [WPGMC]() (**W**eighted **P**air **G**roup **M**ethod with **C**entroid linkage)
- **ward** $^e$  
  - $d(u,v) = \sqrt{\frac{|v| + |s|}{T}d(v,s)^2 + \frac{|v| + |t|}{T}d(v,t)^2 - \frac{|v|}{T}d(s,t)^2}$ 
   - [Ward's minimum variance method](https://en.wikipedia.org/wiki/Ward%27s_method)
   
   
   
$^a)$ $d(u,v)$ Euclidean distanc between $u$ and $v$ clusters.  

$^a)$ $u[i], v[j]$  all points $i$ in cluster $u$ and $j$ in cluster $v$.  

$^b)$ $|u|$ and $|v|$ are the cardinalities of clusters $u$ and $v$ respectiely.

$^c)$ $u, s, v, t$ where cluster $u$ was formed with cluster $s$ and $t$ and $v$ is a remaining cluster in the forest 

$^d)$ $c_s$ and $c_t$ The centroids of clusters $s$ and $t$.
  
$^e)$ $T = |v| + |s| + |t|$  

[Centroid](https://en.wikipedia.org/wiki/Centroid) or geometric center.


| | Arithmetic Mean | Geometric Center |
| :---- | :-------- |:-------- |  
| **Unweighted Pair Group** | average | centroid |
| **Weighted Pair Group**  | weighted | median |


In [20]:
df.iloc[[6,7,18,19], :]

ca_skew  ga_skew  ua_skew  ug_skew  uc_skew  cg_skew
Dengue_MT862858   907.99  -114.88  1841.00  1955.26   952.23  1021.56
Dengue_MT862893   917.53  -124.85  1845.74  1969.50   947.53  1040.92
Zika_AY632535    1381.19  -283.89  1147.27  1432.18  -237.85  1663.35
Zika_MN101548    1352.11  -295.92  1152.43  1444.86  -202.89  1644.59

[Time Complexity](http://en.wikipedia.org/wiki/Time_complexity) (Big O Notaion)  
- $O(n^2)$
 - single  
 - complete  
 - average  
 - weighted  
 - ward
- $O(n^3)$
 - centroid  
 - median  


## 5.3: Cophenetic correlation

You can use the [cophenet](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.cophenet.html) function to calculate the cophenetic coefficent. The [cophenetic correlation](https://en.wikipedia.org/wiki/Cophenetic_correlation) coefficient is a measure of how faithfully a dendrogram preserves the pairwise distances between the original unmodeled data points.

In [21]:
from scipy.cluster.hierarchy import cophenet
from scipy.spatial.distance import pdist

c, coph_dists = cophenet(linkage_matrix, pdist(mat, metric='euclidean'))
c

0.9134217812884088

In [ ]:
#coph_dists
#pdist(mat, metric='euclidean')

In [22]:
import operator
cophenetic_results = {}
methods = ('single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward')
for m in methods:    
    linkage_matrix = linkage(mat, method = m, metric='euclidean', optimal_ordering=True)
    #for x in range(len(linkage_matrix)):
    #    print(linkage_matrix[x])
    c, coph_dists = cophenet(linkage_matrix, pdist(mat))
    cophenetic_results[m] = c
    #print(c)
    #print('======================================================')
print(max(cophenetic_results.items(), key=operator.itemgetter(1))[0])
print(cophenetic_results)

single
{'single': 0.9134217812884088, 'complete': 0.9019088064307287, 'average': 0.9107636361802532, 'weighted': 0.9077800575926043, 'centroid': 0.9106318070979335, 'median': 0.907864065075456, 'ward': 0.8872728478312205}


# Task 6: Dendrogram

In [ ]:
linkage_matrix = linkage(mat, method = 'single', metric='euclidean', optimal_ordering=True)
dendrogram(linkage_matrix)
plt.show()
plt.close()

In [ ]:
linkage_matrix = linkage(mat, method = 'single', metric='euclidean', optimal_ordering=False)
dendrogram(linkage_matrix)
plt.show()
plt.close()

In [ ]:
orientations = ('top', 'bottom', 'left', 'right')
linkage_matrix = linkage(mat, method = 'single', metric='euclidean', optimal_ordering=False)
for orien in orientations:
    dendrogram(linkage_matrix, orientation = orien)
    plt.show()
    plt.close()

In [ ]:
linkage_matrix = linkage(mat, method = 'single', metric='euclidean', optimal_ordering=True)
dendrogram(linkage_matrix, labels=virus_names[:])
plt.show()
plt.close()

In [ ]:
linkage_matrix = linkage(mat, method = 'single', metric='euclidean', optimal_ordering=True)
dendrogram(linkage_matrix, labels=virus_names[:], leaf_rotation=90, leaf_font_size=10)
plt.show()
plt.close()

In [ ]:
plt.figure(figsize=(10, 12))
dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=6230)
plt.show()    
plt.close()

In [ ]:
0.7*max(linkage_matrix[:,2])

**Line style and color options**  
```python
linestyles = ('solid', 'dashed', 'dashdot', 'dotted')
colors_abbreviation = ('b', 'g', 'r', 'c', 'm', 'y', 'k', 'w')
colors_names = ('blue', 'orange', 'green', 'red', 'purple', 'brown', 'pink', 'gray', 'olive', 'cyan')
colors_hexa = ('#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf')
```
*You can read more about matplotlib colors in the [documentation](https://matplotlib.org/api/colors_api.html#matplotlib.colors.Colormap).*

In [ ]:
plt.figure(figsize=(10, 12))
plt.title('20 Viruses Hierarchical \n Clustering Dendrogram', fontsize=16)
plt.xlabel('Euclidean Distance', fontsize=14)
plt.ylabel('The length of the cluster in this direction has no meaning what so ever!', fontsize=14)
dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=0.2*max(linkage_matrix[:,2]))
plt.vlines(x=2000, ymin=1.0, ymax=200., linestyles='dashdot', color='r', lw=2)
plt.vlines(x=[4000,2000, 1000], ymin=[0, 2, 8], ymax=198., linestyles='dashdot', color='#7f7f7f')
#plt.hlines(y=2000, xmin=1.0, xmax=200., linestyles='dashdot', color='r')
#plt.hlines(y=[4000,2000, 1000], xmin=[0, 2, 8], xmax=198., linestyles='dashdot', label='Multiple Lines', color='#7f7f7f')
plt.text(2000, 190, '2000', size=12, ha='center', va='center')
plt.show()    
plt.close()

# Task 7: Dendrogram

 - Riboviria  
     - Coronaviridae family,  
      - Corona 
      - SARS  
     - Flaviviridae family,  
        - Dengue 
        - Zika 
        - West Nile  
     - Picornaviridae  
        - Enterovirus  
     - Retroviridae  
        - HIV  
| | Virus Count | Strain Count |
| :---- | :--------: |:--------: |  
| *Coronaviridae* | **2** | **10** |
| *Flaviviridae*  | **3** | **5** |
| *Picornaviridae* | **1** | **5** |
| *Retroviridae*  | **1** | **1** |
|  |
| **Total**  | **7** | **20** |


In [ ]:
print(cophenetic_results)

In [ ]:
print("{:.3f}".format(cophenetic_results['single'] - cophenetic_results['ward']))

In [ ]:
linkage_matrix = linkage(mat, method = 'single', metric='euclidean', optimal_ordering=False)
plt.figure(figsize=(10, 12))
plt.title('Hierarchical Clustering \n Single Method \n', fontsize=16)
dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=0.12*max(linkage_matrix[:,2]))
plt.vlines(x=[2000, 4000, 9000], ymin=[0, 0, 0], ymax=200., linestyles='dashdot', color='#7f7f7f')
plt.text(2000, 190, '2000', size=12, ha='center', va='center')
plt.text(4000, 190, '4000', size=12, ha='center', va='center')
plt.text(9000, 190, '9000', size=12, ha='center', va='center')
plt.show()    

In [ ]:
linkage_matrix = linkage(mat, method = 'complete', metric='euclidean', optimal_ordering=False)
plt.figure(figsize=(10, 12))
plt.title('Hierarchical Clustering \n Complete Method \n', fontsize=16)
dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=0.12*max(linkage_matrix[:,2]))
plt.vlines(x=[5000, 11000, 20000], ymin=[0, 0, 0], ymax=200., linestyles='dashdot', color='#7f7f7f')
plt.text(5000, 190, '5000', size=12, ha='center', va='center')
plt.text(11000, 190, '11000', size=12, ha='center', va='center')
plt.text(20000, 190, '20000', size=12, ha='center', va='center')

plt.show()    

In [ ]:
linkage_matrix = linkage(mat, method = 'ward', metric='euclidean', optimal_ordering=False)
plt.figure(figsize=(10, 12))
plt.title('Hierarchical Clustering \n Ward\'s Method \n', fontsize=16)
dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=0.12*max(linkage_matrix[:,2]))
plt.vlines(x=[5000, 13000, 39000], ymin=[0, 0, 0], ymax=200., linestyles='dashdot', color='#7f7f7f')
plt.text(5000, 190, '5000', size=12, ha='center', va='center')
plt.text(11000, 190, '11000', size=12, ha='center', va='center')
plt.text(39000, 190, '39000', size=12, ha='center', va='center')

plt.show()    

In [ ]:
linkage_matrix = linkage(mat[:,[0,3]], method = 'single', metric='euclidean', optimal_ordering=False)
plt.figure(figsize=(10, 12))
plt.title('Hierarchical Clustering \n Single Method \n ', fontsize=16)
dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=0.12*max(linkage_matrix[:,2]))
plt.show()    

In [ ]:
methods = ('single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward')
for m in methods:    
    linkage_matrix = linkage(mat, method = m, metric='euclidean', optimal_ordering=False)
    y = 0
    plt.figure(figsize=(8, 9))
    plt.title('Hierarchical Clustering Dendrogram using ' + str(m) + ' Method', fontsize=16)
    dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=0.2*max(linkage_matrix[:,2]))
    plt.show()   

In [ ]:
methods = ('single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward')
for m in methods:    
    linkage_matrix = linkage(mat2, method = m, metric='euclidean', optimal_ordering=False)
    y = 0
    plt.figure(figsize=(8, 9))
    plt.title('Hierarchical Clustering Dendrogram using ' + str(m) + ' Method', fontsize=16)
    dendrogram(linkage_matrix, labels=virus_names[:], orientation="left", color_threshold=0.2*max(linkage_matrix[:,2]))
    plt.show()    

<table><tr><td><img src='images/dendogram_H.png'></td><td><img src='images/dendogram.png'></td></tr></table>  